buildplot

LamAdr opened the pull request:
ok, I hope this is your conventional pull request and that you haven't been spammed by my unfortunate earlier attempts.

vincentarelbundock commented:
Thanks, this is great!

I pushed a change to your branch and added some new tests. Some of them fail, which illustrates needed improvements in the function. 2 things in particular:

+ [ ] When `condition` is a single string, wrap it in a list so that the rest works without changes.
+ [ ] Check to make sure that the values of the dict conform to the types of the columns. In one of your tests, we have `Department` as a numeric, but the dict values should be strings for this, because that is what the column contains. In Polars, I think you can do something like: `dat["column_name"].is_numeric()`
+ [ ] If the dict value is None, then you should set the values automatically as if the user had just supplied a string (100 points on the range, unique values, or Tukey's 5 numbers.)

LamAdr commented:
I made corrections

vincentarelbundock commented:
@LamAdr Alright, this looks great. You can move on to steps 2 and 3 (with matplotlib) whenever you are ready.

vincentarelbundock commented:
I'm not in a special rush, but just to get an idea, when do you think you'll have time to make progress on this @LamAdr ?

Thanks!

LamAdr commented:
@vincentarelbundock
Yeah sorry the past weeks have been busy. I should be able to submit steps 2&3 before the end of this week.

vincentarelbundock commented:
No worries at all. It was just to think about my own calendar...

Thanks!

vincentarelbundock commented:
This looks really great, thanks! I pushed a minor commit with:

* Return the `matplotlib` object instead of calling `.show()` internally. This allows users to use `.savefig()` or other methods instead of drawing the plot immediately.
* Fixed the tests

Next steps:

+ [ ] `draw` argument. `True` is default. When `draw=False`, the `plot_predictions()` function returns the custom data frame instead of the plot.
+ [ ] New arguments which are easy to push directly from `plot_predictions()` to `predictions()`: `transform`, `vcov`, `conf_level`. Put them in the same order as in `R`: https://marginaleffects.com/articles/reference/plot_predictions.html#usage
+ [ ] New `by` argument. This one uses essentially the same plotting code, but it bypasses the `condition` data frame building step. Instead, it feeds the `newdata` content and uses the `by` argument in `predictions` to make predictions by subgroups. 
  - [ ] `assert`: When `by` is used, the user must supply `newdata`
  - [ ] When `by` is used, the user must not supply `condition`.

LamAdr commented:
Thank you for fixing my code @vincentarelbundock.
I am not sure about the third point of this next task. Is it a departure from the `R` package where, if I understand correctly, one can use `by` without `newdata`, but not the converse?

vincentarelbundock commented:
> Thank you for fixing my code @vincentarelbundock. I am not sure about the third point of this next task. Is it a departure from the `R` package where, if I understand correctly, one can use `by` without `newdata`, but not the converse?

Ah yes, sorry I made a mistake. If `by` is used and `newdata` is not, then we extract the original dataset used to fit the model with `get_modeldata()`. Good catch!

LamAdr commented:
Hi @vincentarelbundock, is this the right logic?
Should I add the other arguments (type, wts, points, rug, gray)?
Also would you like me to try to horizontally distance dots that belong to the same category with respect to the x-axis but of different colors?

vincentarelbundock commented on marginaleffects/plot.py:
```diff
@@ -0,0 +1,193 @@
+            by = [by]
+
+        if newdata is not None:
+            dt = predictions(model, newdata=newdata, conf_level=conf_level, vcov=vcov, transform=transform)
```
is there a `by` argument missing here?

vincentarelbundock commented on marginaleffects/plot.py:
```diff
@@ -0,0 +1,193 @@
+        dt = dt.sort(by[0])
+
+    if condition is not None:
+        built_dt = build_plot(model, condition)
```
maybe rename `build_plot()` to something more explicit to indicate that this function is used only with the `condition` argument.

vincentarelbundock commented on marginaleffects/plot.py:
```diff
@@ -0,0 +1,193 @@
+
+            common_plot(color_dt, by[0], x_type, fig=fig, label=by_1val)
+
+        ax.legend(loc='center left', bbox_to_anchor=(1, 0.5), title=by[1], fontsize=titles_fontsize, title_fontsize=titles_fontsize)
```
I tried a couple plots and sometimes the legend seems to go out of bounds of the plot. But this may be because of my setup, which is weird (ubuntu via WSL on windows).

LamAdr replied:
I'm not sure what you mean. Like this ?

![Screenshot from 2023-08-30 06-33-33](https://github.com/vincentarelbundock/pymarginaleffects/assets/102169875/8fc902d0-8e7f-4fbc-91f8-69a29a917302)

vincentarelbundock replied:
Yes

LamAdr replied:
Does it displays correctly when the window is expanded?
Do you want me to shrink the plot so that everything appears in the small window?

vincentarelbundock replied:
Can't check now, but this is low priority. Let's keep things as is for now.

vincentarelbundock commented:
Yep, thanks. Overall the logic looks great. I left a few minor comments in the code but think we can generally move forward.

Yes for `type` and `wts`. No for the other ones. I want to get `plot_predictions()`, `plot_comparisons()`, and `plot_slopes()` all working as expected and well tested before doing any of the aesthetic stuff.

I think the other two functions will be much more easy to implement because we can re-use a bunch of the code you've already written.

LamAdr commented:
Okay, just to let you know, I am doing a student competition this week, you probably won't hear from me until next one.

LamAdr commented:
Minor changes. It seems that `predictions` no longer outputs the columns without `by`, so I added that argument when used with `condition`. It also adds a `null` row when `by` is used together with `newdata` so I just dropped that row.
@vincentarelbundock

LamAdr commented:
Hello @vincentarelbundock, please review when you have time.
I haven't been able to debug the 2nd test of `test_plot_comparisons` yet. The one with `contrast` = `"sd"`.
Still working on this.

vincentarelbundock commented:
All of this looks really great. Thanks so much!

I think once these three steps are completed, we'll be able to merge the branch:

1. Fix the bug that you noted.
2. Add documentation to the three top level `plot_*()` functions. You can go to the `marginaleffects` for R package and copy over most of it verbatim, with only minor adjustment for Python vs. R vocabulary. Ex: https://marginaleffects.com/articles/reference/plot_predictions.html
3. Create some "snapshot tests" using `pytest` or some other framework that you like. Basically, we save images for a few basic plots, compare them to the R output to make sure that they are correct. Then, we include `pytest` tests that will compare the output to our known/saved "gold standard" snapshots. 

Does that work for you?

LamAdr commented:
In step 3, do we compare the images or the underlying dataframe?
Otherwise that absolutely works.

vincentarelbundock commented:
Normally we compare the actual images, with some tolerance, ex: 97% of pixels need to be identical.

I don't know the frameworks in Python, but I'm guessing there will be one in `pytest` that is more popular or modern than the others. Here's a link to the R package that many people use. You can use the vocabulary on that page to google for a python equivalent (or ask ChatGPT): https://vdiffr.r-lib.org/

LamAdr commented:
@vincentarelbundock is this right?

vincentarelbundock commented:
have not tried it, but it looks right to me. `newdata` is sometimes a 1 row dataset (ex: when we call `newdata="mean"`), so we don't want to take the standard deviation of that. Instead, we need to take the standard deviation of the column in the original dataset (`modeldata`).

vincentarelbundock commented:
Ideally you would run a similar command in R to make sure you get the same plot.

LamAdr commented:
@vincentarelbundock 
Documentation is ready for review.
`plot_predictions` and `plot_comparisons` tests are working for me on Ubuntu and Windows with this level of tolerance.
`plot_slopes` tests are not because the standard errors are noisy. Is this difference with `comparisons` expected?

LamAdr commented:
Okay, I can hear you from here telling me to answer my own question by comparing with `R` output, and honestly that's genius so I will do that and report back.

vincentarelbundock commented:
Haha nice. I'll.wait for that before reviewing

LamAdr commented:
@vincentarelbundock 
I am not sure what to do now.

As you can see, the 3rd test of plot_slopes produces a figure with noisy errors.
![Figure_3](https://github.com/vincentarelbundock/pymarginaleffects/assets/102169875/c0ced664-5665-4f6f-850c-62ee0b5afb50)
The errors on the corresponding R figure are smooth.


Also, when I run the python test on Windows I get this difference :
![Figure_3-failed-diff](https://github.com/vincentarelbundock/pymarginaleffects/assets/102169875/7b3f8383-6989-4a2a-b317-73b98a2e1b47)
As you can see, not only the errors are differing, but also the estimated functions.
The only packages differences I have are for :
- exceptiongroup
- iniconfig
- pluggy
- pytest
- tomli
- colorama

I'm guessing the two observations are not independent. Could there be an issue in the error computation in python? Do you want me to investigate that? Or do you have another suggestion?

Many thanks

vincentarelbundock commented:
Comment of this test but level it in the file, then open a new Github issue about it. Then, let me know if you think this is ready to merge.

I will investigate after merge.

LamAdr commented:
@vincentarelbundock What do you mean by "level it in the file"?

vincentarelbundock commented:
Leave it in (typo)

LamAdr commented:
I think this is ready to merge

vincentarelbundock commented:
Merged! Thanks so much for this. I think it's a really cool feature and you did amazing work. I'm *__very__* happy with the outcome.
